private/pkg/manifest/storage.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manifest
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"github.com/bufbuild/buf/private/pkg/storage/storageutil"
+	"go.uber.org/multierr"
+)
+
+// manifestBucket is a storage.ReadBucket implementation from a manifest an an
+// array of blobs.
+type manifestBucket struct {
+	manifest   Manifest
+	pathToBlob map[string]Blob
+}
+
+type manifestBucketObject struct {
+	path string
+	file io.ReadCloser
+}
+
+func (o *manifestBucketObject) Path() string               { return o.path }
+func (o *manifestBucketObject) ExternalPath() string       { return o.path }
+func (o *manifestBucketObject) Read(p []byte) (int, error) { return o.file.Read(p) }
+func (o *manifestBucketObject) Close() error               { return o.file.Close() }
+
+// NewFromBucket creates a manifest from a storage bucket, with all its digests
+// in DigestTypeShake256.
+func NewFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := New()
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	if walkErr := bucket.Walk(ctx, "", func(info storage.ObjectInfo) (retErr error) {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
+		if err != nil {
+			return err
+		}
+		defer func() { retErr = multierr.Append(retErr, obj.Close()) }()
+		digest, err := digester.Digest(obj)
+		if err != nil {
+			return err
+		}
+		if err := m.AddEntry(path, *digest); err != nil {
+			return err
+		}
+		return nil
+	}); walkErr != nil {
+		return nil, walkErr
+	}
+	return m, nil
+}
+
+// NewBucket takes a manifest and an array of blobs and builds a readable storage bucket that
+// contains the files in the manifest.
+func NewBucket(m Manifest, blobs []Blob) (storage.ReadBucket, error) {
```

Comment:
I feel very strongly we should receive a map here instead
```suggestion
func NewBucket(m Manifest, blobs map[string]Blob) (storage.ReadBucket, error) {
```

That way we don't need to do the checking of duplicate blobs (and possible different content) for the same digest.

Reply:
Although, on second thought, in here we'd need to still check that key and value matches:

```
{
  "shake256:abc123": {"digest": "shake256:abc123", "content": "..."}
  |----- x -------|             |----- x -------|
}
```

So let's keep it as array I guess 😢 

For more background, I was thinking of converting everything to just `map[digest]content` and *that* would be the bag of blobs. But that has two downsides:

- A manifest couldn't any longer be represented as a Blob (or a weird map of 1 entry?)
- A digest(kind, bytes) as map key is not fun for proto, and we'd need to always use the string representation of it (would have this been solved by using multihash, where everything is just []byte?)

Reply:
I totally understand the want for a map; it's the natural way of declaring a set in Go. Precisely the type we wish to describe a bag of blobs.

multihash would solve one problem in that it's not a composite of two values, but []byte isn't a valid map key type. We would need to different abstraction for a set.